Plug-Ins/NovaTools/System Editor/SystemWindowController.swift

```diff
@@ -0,0 +1,335 @@
+            _ = try reader.read() as Int16
+        }
+        stellarList = try (0..<16).map { _ in
+            Int(try reader.read() as Int16)
+        }.map { stellarId in
+            if 128...2175 ~= stellarId {
+                manager.findResource(type: ResourceType("spöb"), id: stellarId, currentDocumentOnly: false)
+            } else {
+                nil
+            }
+        }
+     }
+
+    private func read(stellar: Resource) {
+        if let view = stellarViews[stellar.id], view.resource == stellar {
+            try? view.read()
+        } else if let view = StellarView(stellar, manager: manager, isEnabled: stellar.document == manager.document) {
+            stellarViews[stellar.id] = view
+        }
+    }
+
+    @IBAction func zoomIn(_ sender: Any) {
+        systemView.zoomIn(sender)
+    }
+
+    @IBAction func zoomOut(_ sender: Any) {
+        systemView.zoomOut(sender)
+    }
+
+    @IBAction func copy(_ sender: Any) {
+        let pb = NSPasteboard(name: .general)
+        pb.declareTypes([.RFResource], owner: nil)
+        pb.writeObjects(selectedStellars)
+    }
+
+    @IBAction func paste(_ sender: Any) {
+        // Forward to the document
+        manager.document?.perform(#selector(paste(_:)), with: sender)
+    }
+
+    func createStellar(position: NSPoint = .zero, navDefault: Int? = nil) {
+        let destinationNavDefault = navDefault ?? stellarList.firstIndex(of: nil) ?? -1
+        // Make sure there's room for another stellar in the system, and don't allow overwriting an existing navDefault
+        if !(0...15 ~= destinationNavDefault) || stellarList[destinationNavDefault] != nil {
+            return
+        }
+        manager.createResource(type: ResourceType("spöb"), id: nil) { [weak self] stellar in
+            guard let self else { return }
+
+            // Construct the minimum data required
+            let writer = BinaryDataWriter()
+            writer.write(Int16(position.x.rounded()))
+            writer.write(Int16(position.y.rounded()))
+            for _ in 0..<16 {
+                writer.write(Int16(-1))
+            }
+            // Allow the DataChanged notification to create the view
+            stellar.data = writer.data
+            // Add the stellar to our navdefaults
+            stellarList[destinationNavDefault] = stellar
+
+            // Save the nav defaults
+            let systWriter = BinaryDataWriter()
+            // Make sure there's enough data in the resource to save nav defaults; otherwise initialize default data up to the end of the NavDefaults list
+            if resource.data.count < 2 * 2 {
+                systWriter.write(Int16(0)) // xPos/yPos = 0
+                systWriter.write(Int16(0))
+            }
+            if resource.data.count < (2 + 16 + 16) * 2 {
+                for _ in resource.data.count..<32 {
+                    systWriter.write(Int16(-1)) // no hyperlinks or navdefaults
+                }
+            } else {
```

Comment:
This doesn't quite look right, I think the if statements need to be combined. If the first one is true but the second one if false, the work done in the first will just get replaced anyway.
```suggestion
            if resource.data.count < (2 + 16 + 16) * 2 {
                systWriter.write(Int16(0)) // xPos/yPos = 0
                systWriter.write(Int16(0))
                for _ in resource.data.count..<32 {
                    systWriter.write(Int16(-1)) // no hyperlinks or navdefaults
                }
            } else {
```

Reply:
> This doesn't quite look right

My first implementation was pretty much what you suggested but after testing I changed it to my current approach because there seemed to be (at least) three different states a `sÿst` resource can be in:
1. A full resource with completely written data: happy days, we just update the nav defaults block.
2. A brand new resource with zero bytes of data written (e.g. newly created): this triggers both `if` conditions and gives us the minimum amount of data required to update nav defaults.
3. A newly-created `sÿst` resource with only xPos/yPos and connections: this triggers only the second `if` condition to expand the remaining defaults to -1 so we don't overwrite the existing data on save.

> If the first one is true but the second one if false, the work done in the first will just get replaced anyway.

If the first is true the second will always be true — looking at it again it would be clearer if I put the first `if` block inside the second.

Reply:
Ah yes, but this is changed by your previous PR with reading the nav defaults in the galaxy. I'm just looking at that again now and realise it actually broke adding a new system because the create process isn't writing enough data for the nav defaults. If we fix that in the galaxy editor then it will ensure that the 3rd option above should always have enough data for what you need here. So it should be fine to just do one check for sufficient data, and if there isn't enough then just start from scratch without regard for how much data there actually is.

Reply:
I've just pushed a fix for the syst creation in the galaxy - it now writes the nav defaults too.